comet-java-client/src/main/java/ml/comet/experiment/impl/BaseExperiment.java

```diff
@@ -0,0 +1,899 @@
+                getLogger().error("failed to close connection", e);
+            }
+        }
+
+        // dispose all pending calls
+        if (disposables.size() > 0) {
+            getLogger().warn("{} calls still has not been processed, disposing", disposables.size());
+        }
+        this.disposables.dispose();
+    }
+
+    /**
+     * Sends heartbeat to the server and returns the status response.
+     *
+     * @return the status response of the experiment.
+     */
+    Optional<ExperimentStatusResponse> sendExperimentStatus() {
+        return Optional.ofNullable(validateAndGetExperimentKey()
+                .concatMap(experimentKey -> restApiClient.sendExperimentStatus(experimentKey))
+                .onErrorComplete()
+                .blockingGet());
+    }
+
+    /**
+     * Synchronously loads remote data using provided load function or throws an exception.
+     *
+     * @param loadFunc the function to be applied to load remote data.
+     * @param alias    the data type alias used for logging.
+     * @param <T>      the data type to be returned.
+     * @return the loaded data.
+     */
+    private <T> T loadRemote(final Function<String, Single<T>> loadFunc, String alias) {
+        return validateAndGetExperimentKey()
+                .concatMap(loadFunc)
+                .doOnError(ex -> getLogger().error("Failed to read {} for the experiment, experiment key: {}",
+                        alias, this.experimentKey, ex))
+                .blockingGet();
+    }
+
+    /**
+     * Uses provided function to send request data asynchronously and log received output. Optionally, can use
+     * provided {@link Action} handler to notify about completion of the operation.
+     *
+     * @param func       the function to be invoked to send request data.
+     * @param request    the request data object.
+     * @param onComplete the optional {@link Action} to be notified the operation completes either
+     *                   successfully or erroneously.
+     * @param <T>        the type of the request data object.
+     */
+    private <T> void sendAsynchronously(final BiFunction<T, String, Single<LogDataResponse>> func,
+                                        final T request, final Action onComplete) {
+        Single<LogDataResponse> single = validateAndGetExperimentKey()
+                .subscribeOn(Schedulers.io())
+                .concatMap(experimentKey -> func.apply(request, experimentKey));
+
+        // register notification action if provided
+        if (onComplete != null) {
+            single = single.doFinally(onComplete);
+        }
+
+        // subscribe to receive operation results
+        single
+                .observeOn(Schedulers.single())
+                .subscribe(
+                        (logDataResponse) -> DataResponseLogger.checkAndLog(logDataResponse, getLogger(), request),
+                        (throwable) -> getLogger().error("failed to log {}", request, throwable),
```

Comment:
If the backend is down and the user tries to send 10000 metrics value, will they get their console filled with 10000 log message saying `failed to log`?

Reply:
Yes. 
We can create ticket for this as future improvements with details what is expected.

Reply:
Looks like it's the same behavior as before, right? https://github.com/comet-ml/comet-java-sdk/pull/37/files#diff-380c0a24ffacf3b646e64f8bfbd9840582db49fdb2ad732798905dfbf247a455L118

Reply:
Sorry, I didn't get you last comment. Can we communicate with full statements rather than hint question?

Reply:
I was trying to see if there was a behavior change between the code on master and this PR. From my understanding, the log message is different but with the current release Java SDK, if the backend goes down and the user tries to send 10000 metrics, they will get 10000 error message. Could you confirm that?

Reply:
Yes, I've already confirmed this. If user going to send billion metrics he/she will get the same amount of errors logged in case the server is down.